### PR TITLE
yarn: fix to not use npm to install

### PR DESCRIPTION
**Pull request by SimenB**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This doesn't pass it's test, and I'm unable to use it locally after install. There seems to be some issue  with the symlink. Instead of following it it's just prepended to the current path. So if I try to run yarn from a deep path:

```sh-session
$ ls -la /usr/local/Cellar/yarn/0.21.3/bin
total 16
drwxr-xr-x  4 simen  admin  136 Mar 31 08:21 .
drwxr-xr-x  8 simen  admin  272 Mar 31 08:21 ..
lrwxr-xr-x  1 simen  admin   19 Mar 31 08:21 yarn -> ../libexec/bin/yarn
lrwxr-xr-x  1 simen  admin   22 Mar 31 08:21 yarnpkg -> ../libexec/bin/yarnpkg
$ ls -la /usr/local/Cellar/yarn/0.21.3/libexec/bin
total 32
drwxr-xr-x   7 simen  admin   238 Feb 27 16:36 .
drwxr-xr-x  10 simen  admin   340 Mar 31 08:21 ..
drwxr-xr-x   4 simen  admin   136 Feb 27 16:36 node-gyp-bin
-rwxr-xr-x   1 simen  admin   906 Feb 27 16:29 yarn
-rw-r--r--   1 simen  admin    34 Feb 27 16:29 yarn.cmd
-rwxr-xr-x   1 simen  admin  1874 Feb 27 16:29 yarn.js
-rwxr-xr-x   1 simen  admin    42 Feb 27 16:29 yarnpkg
$ pwd
/Users/simen
$ /usr/local/Cellar/yarn/0.21.3/libexec/bin/yarn --version
0.21.3
$ /usr/local/Cellar/yarn/0.21.3/bin/yarn --version
module.js:472
    throw err;
    ^

Error: Cannot find module '/Users/libexec/bin/yarn.js'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.runMain (module.js:605:10)
    at run (bootstrap_node.js:423:7)
    at startup (bootstrap_node.js:147:9)
    at bootstrap_node.js:538:3
```

Why does this happen? `brew test` fails because of this. Might be that the old npm solution put up some extra magic symlinks, but I can't find them, beyond the fact that old `libexec/bin` also was a symlink.

(this is using the old install)
```sh-session
$ ls -la /usr/local/Cellar/yarn/0.21.3/libexec/bin
total 16
drwxr-xr-x  4 simen  admin  136 Mar 31 08:33 .
drwxr-xr-x  5 simen  admin  170 Mar 31 08:33 ..
lrwxr-xr-x  1 simen  admin   36 Mar 31 08:33 yarn -> ../lib/node_modules/yarn/bin/yarn.js
lrwxr-xr-x  1 simen  admin   36 Mar 31 08:33 yarnpkg -> ../lib/node_modules/yarn/bin/yarn.js
```

/cc @Daniel15

### CODE REVIEW

**Comment by SimenB** [comment]
Ah, figured it out! The clue was in the last one. The old links pointed directly to the `yarn.js` file, not to the incuded `yarn` and `yarnpkg`. It works now 😄 Sorry about the noise in the OP, but I'll keep it for reference.

**Comment by SimenB** [comment]
This might be partly #6936 /cc @torifat

**Review by ilovezfs — CHANGES_REQUESTED** [review]
This has 42,055 installs in the last 30 days. I'd prefer we do such "experiments" on something a bit more obscure than this.

**Comment by SimenB** [comment]
Not sure how this is "an experiment"? It exchanges an unsupported way of installing yarn with a supported one. Aren't the tests to be trusted?

And the fact that npm is used to install broke the installation from brew for me yesterday since I got a new npm from a newer node. So even though the tarball itself was unchanged, I was unable to use a brew-installed yarn without resetting brew to a place in history with another node version

**Comment by ilovezfs** [comment]
>It exchanges an unsupported way of installing yarn with a supported one.

That could describe a huge number of formulae, so not really convincing.

>And the fact that npm is used to install broke the installation from brew for me yesterday since I got a new npm from a newer node. So even though the tarball itself was unchanged, I was unable to use a brew-installed yarn without resetting brew to a place in history with another node version.

Not sure I understand what that means but it's not really relevant to this specific PR and sounds like something you should open an issue about.

File: Formula/yarn.rb
Comment (MikeMcQuaid) [inline Formula/yarn.rb]: Pretty much my only objection here is if they add additional commands that we now need to remember to update/manually add them. Thoughts?

Comment (SimenB) [inline Formula/yarn.rb]: Those are extracted from the `bin` object in package.json. Is there some easy way to use something like `jq` to extract it?

https://github.com/yarnpkg/yarn/blob/76c9fc181e819d76d6f0833f57313c646b9b5406/package.json#L82-L85

Comment (MikeMcQuaid) [inline Formula/yarn.rb]: You could use Ruby's JSON support to do so but it worries me a little if we get to the point that we're parsing these files rather than letting `npm` or similar do so...

Comment (ilovezfs) [inline Formula/yarn.rb]: Note this is handled without any additional fuss by the current formula.

Comment (SimenB) [inline Formula/yarn.rb]: That said, I doubt they'll add new ones. `yarn` is the name of the thing, but since some debian distros already have a global executable called `yarnpkg`, that's added as an alias. @Daniel15 can probably comment

Comment (MikeMcQuaid) [inline Formula/yarn.rb]: In general @Daniel15 having very strong feelings on this PR either way would be valued input.

Comment (Daniel15) [inline Formula/yarn.rb]: I doubt we'll add additional executables other than `yarn` and `yarnpkg` so I think it's fine to just hard-code those two. I don't know enough about Homebrew to have strong feelings either way. 

The executables are in a `bin` directory, could Homebrew just link all the extensionless files in the `bin` directory? https://github.com/yarnpkg/yarn/tree/master/bin

Comment (SimenB) [inline Formula/yarn.rb]: @Daniel15 That's what I had at the start, but the `yarn` executable in there was unable to properly follow symlinks (or it followed symlinks it shouldn't, not sure?), so I ended in the situation I describe in the OP

Comment (SimenB) [inline Formula/yarn.rb]:
```sh-session
$ readlink /usr/local/bin/yarn
../Cellar/yarn/0.21.3/bin/yarn
$ realpath /usr/local/bin/yarn
/usr/local/Cellar/yarn/0.21.3/libexec/bin/yarn.js
```

Can `realpath` be trusted to exist on the platforms that runs that shellscript?

Although `yarnpkg` just runs `yarn.js`, it doesn't do the stuff `yarn` does. Is that by design?

**Comment by SimenB** [comment]
@MikeMcQuaid Changed the symlinking to read paths and binaries from package.json

**Comment by ilovezfs** [comment]
That seems overly complex given what Daniel said above.

**Comment by SimenB** [comment]
Like I replied there, it doesn't work. That was the whole trouble I had in the OP, since I just symlinked the whole bin folder. In addition to non-working scripts, I also got extra files linked. I can filter those out, or just use a whitelist, but then it isn't really any less complex, IMO 😄

**Comment by Daniel15** [comment]
The code in this PR seems reasonable to me. It's not overly complex.

**Comment by ilovezfs** [comment]
I would hard code the list not parse json.

**Comment by Daniel15** [comment]
The code in this PR seems reasonable to me. It's up to the Homebrew maintainers though :)

File: Formula/yarn.rb
Comment (ilovezfs) [inline Formula/yarn.rb]: If we're going down this road of parsing these json packaging files ourselves, which I don't think is advisable, it would belong in the brew backend with a DSL, not proliferated throughout formulae.

**Comment by Daniel15** [comment]
Yeah hardcoding `yarn` and `yarnpkg` would be fine too. We do that in Yarn itself for things like the Debian package.

**Comment by ilovezfs** [comment]
>Yeah hardcoding yarn and yarnpkg would be fine too. We do that in Yarn itself for things like the Debian package.

This.

File: Formula/yarn.rb
Comment (SimenB) [inline Formula/yarn.rb]: That I can agree on, I was actually thinking the exact same thing 😄

**Comment by SimenB** [comment]
> I would hard code the list not parse json.

That's what I had. And again, the extensionless files in `/bin` doesn't work, so that would require some work over in Yarn as well.
And the old solution symlinked the symlinks I'm now creating manually.

If it's ok to hardcode to `bin/yarn.js` I can do that? That is what I had before though, see https://github.com/Homebrew/homebrew-core/pull/11881#discussion_r109109418

**Comment by ilovezfs** [comment]
@SimenB Yes, but now we have new information since Daniel said, "I doubt we'll add additional executables other than yarn and yarnpkg so I think it's fine to just hard-code those two."

**Comment by Daniel15** [comment]
@SimenB Here's the Debian package script, for reference: https://github.com/yarnpkg/yarn/blob/fda53e97b07d32982d9ada611588d4baedc7ff51/scripts/build-deb.sh#L57-L59. In the Debian package, we symlink `/usr/bin/yarn` to `/usr/share/yarn/bin/yarn`, where `/usr/share/yarn/bin/yarn` is `bin/yarn` from the Yarn repo/tarball. Couldn't you do the same thing?

**Comment by SimenB** [comment]
Cool, I'll revert the changes I did today 😄 

@Daniel15 No, `readlink` returns a relative path, and while yarnpkg/yarn#2441 fixes it for GNU `readlink`s, on Mac you get `illegal option -- f`. If `yarn` can use `realpath` instead of `readlink` that would fix it, but I'm not sure how viable that is.

**Comment by Daniel15** [comment]
Oh yeah, I always forget that Mac uses BSD tools rather than GNU tools 😛 

If you like, you could modify Yarn's `bin/yarn` shell script to use the right command based on the environment? It already has some hacks for Cygwin/MinGW on Windows.

**Comment by SimenB** [comment]
I can try sending a PR for it sure, but then we'd have to hold off on this PR until a new (stable) release of yarn. No biggie of course 😄

**Comment by ilovezfs** [comment]
>If you like, you could modify Yarn's bin/yarn shell script to use the right command based on the environment? It already has some hacks for Cygwin/MinGW on Windows.

Yeah it would be great to have a BSD friendly shell script. Note that `realpath` itself isn't a command on macOS, but only available as
```
char * realpath(const char *restrict file_name, char *restrict resolved_name);
```
(unless you've installed `coreutils`, but that's a pretty heavy dependency to add, which wouldn't pass the smell test.)

**Comment by ilovezfs** [comment]
>I can try sending a PR for it sure, but then we'd have to hold off on this PR until a new (stable) release of yarn.

Assuming upstream indicated an intent to accept that PR, we could always apply it as a patch here.

**Comment by SimenB** [comment]
We can merge this linking yarn.js and if we can make the `bin/yarn` behave better update what we link?

**Comment by ilovezfs** [comment]
>We can merge this linking yarn.js and if we can make the bin/yarn behave better update what we link?

Yes, whatever works properly without adding unnecessary dependencies! :)

**Comment by SimenB** [comment]
Ok, removed the changes involving package.json

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
This seems fine to me as long as it's well tested before it goes in.

**Comment by SimenB** [comment]
I have tested it on my personal Mac and the one I use at work. Is there any more testing I can do?

**Comment by ilovezfs** [comment]
@SimenB You could try it with https://github.com/Homebrew/homebrew-core/pull/11991 which wants to use `yarn` as a dependency, and with `lumo`, which already does https://github.com/Homebrew/homebrew-core/blob/master/Formula/lumo.rb#L17 but won't have been tested by CI since it's build-time only.

**Comment by SimenB** [comment]
Lumo worked fine. Grafana had some issue. I don't think it's the installed yarn's fault, though, as I get the same error with the yarn installed from master.

```
$ brew install grafana
==> Downloading https://homebrew.bintray.com/bottles/grafana-4.2.0.sierra.bottle.tar.gz

curl: (22) The requested URL returned error: 401 Unauthorized
Error: Failed to download resource "grafana"
Download failed: https://homebrew.bintray.com/bottles/grafana-4.2.0.sierra.bottle.tar.gz
Warning: Bottle installation failed: building from source.
==> Using the sandbox
==> Downloading https://github.com/grafana/grafana/archive/v4.2.0.tar.gz
==> Downloading from https://codeload.github.com/grafana/grafana/tar.gz/v4.2.0
######################################################################## 100.0%
==> go run build.go build
==> yarn install
==> npm install grunt-cli --verbose
==> node_modules/grunt-cli/bin/grunt build
Last 15 lines from /Users/simen/Library/Logs/Homebrew/grafana/04.grunt:

05 04 2017 08:25:24.031:ERROR [launcher]: Cannot start PhantomJS

05 04 2017 08:25:24.034:INFO [launcher]: Trying to start PhantomJS again (2/2).
05 04 2017 08:25:24.950:ERROR [phantomjs.launcher]: 2017-04-05 08:25:24.950 phantomjs[88158:459644] -[_NSXPCDistantObjectWithError performScanWithChannels:ssidList:legacyScanSSID:includeHiddenNetworks:mergeScanResults:maxAge:maxMissCount:maxWakeCount:maxAutoJoinCount:interfaceName:waitForWiFi:waitForBluetooth:token:priority:reply:]: unrecognized selector sent to instance 0x7fe033808f40

05 04 2017 08:25:24.951:ERROR [phantomjs.launcher]: PhantomJS has crashed. Please read the bug reporting guide at
<http://phantomjs.org/bug-reporting.html> and file a bug report.

05 04 2017 08:25:24.953:ERROR [launcher]: Cannot start PhantomJS

05 04 2017 08:25:24.955:ERROR [launcher]: PhantomJS failed 2 times (cannot start). Giving up.
Warning: Task "karma:test" failed. Use --force to continue.

Aborted due to warnings.

READ THIS: http://docs.brew.sh/Troubleshooting.html

These open issues may also help:
Grafana v4.2.0 https://github.com/Homebrew/homebrew-core/pull/11991
```

**Comment by ilovezfs** [comment]
@SimenB cool. Yeah grafana needs that `--force`. See https://github.com/Homebrew/homebrew-core/pull/9918/files#diff-7473b50efa5c9abb0b107de1cb79c8f7R37

**Comment by ilovezfs** [comment]
@SimenB I think you can also make this ` bottle :unneeded` now.

**Comment by SimenB** [comment]
What does `bottle :unneeded` do?

Building grafana with this diff from #11991 

```diff
diff --git i/Formula/grafana.rb w/Formula/grafana.rb
index 49b03993f..0e15a2d14 100644
--- i/Formula/grafana.rb
+++ w/Formula/grafana.rb
@@ -27,7 +27,10 @@ class Grafana < Formula
       system "go", "run", "build.go", "build"
       system "yarn", "install"
       system "npm", "install", "grunt-cli", *Language::Node.local_npm_install_args
-      system "node_modules/grunt-cli/bin/grunt", "build"
+      args = ["build"]
+      # Avoid PhantomJS error "unrecognized selector sent to instance"
+      args << "--force" unless build.bottle?
+      system "node_modules/grunt-cli/bin/grunt", *args
 
       bin.install "bin/grafana-cli"
       bin.install "bin/grafana-server"
```

**Comment by ilovezfs** [comment]
>What does bottle :unneeded do?

It's a replacement for the bottle block, which you'll delete as well.

The users will `brew install yarn`, triggering a download of https://yarnpkg.com/downloads/0.21.3/yarn-v0.21.3.tar.gz instead of a Bintray bottle, and then it will execute your install method.

**Comment by SimenB** [comment]
Grafana passed with `--force`. And added `bottle :unneeded`, thanks!

**Comment by ilovezfs** [comment]
>Grafana passed with --force.

Great!

Do you mind opening an issue in the yarn issue tracker about the script issue with BSD userland?

File: Formula/yarn.rb
Comment (ilovezfs) [inline Formula/yarn.rb]: Since it's `bottle :unneeded` now, please also add
```
mirror "https://github.com/yarnpkg/yarn/releases/download/v0.21.3/yarn-v0.21.3.tar.gz"
```

Comment (Daniel15) [inline Formula/yarn.rb]: What does `mirror` do? The Yarn download URL just redirects to the GitHub one (we use the yarnpkg.com URL so we're not tightly-coupled with Github's release system)

Comment (ilovezfs) [inline Formula/yarn.rb]: Plot foiled! In that case it would only help if yarnpkg.com went down.

I guess we could use `http://distfiles.gentoo.org/distfiles/yarn-v0.21.3.tar.gz` or be mirror-less.

**Comment by ilovezfs** [comment]
@Daniel15 any changes you'd like to see here before we :ship:?

File: Formula/yarn.rb
Comment (ilovezfs) [inline Formula/yarn.rb]: @SimenB let's add
```
revision 1
```

File: Formula/yarn.rb
Comment (Daniel15) [inline Formula/yarn.rb]: You could keep the mirror just in case `yarnpkg.com` goes down, but the Windows Chocolatey package also uses yarnpkg.com and I haven't seen any reports of being unable to hit yarnpkg.com.

**Comment by Daniel15** [comment]
@ilovezfs It seems reasonable to me, but as I mentioned earlier I'm not very familiar with Homebrew (I used it back when I had a Mac, but I switched back to a PC around a year ago).

By the way, do you really love ZFS? 😄

**Comment by ilovezfs** [comment]
>but I switched back to a PC around a year ago

Heresy!

>By the way, do you really love ZFS?

Yes: https://openzfsonosx.org/

**Comment by MikeMcQuaid** [comment]
@Daniel15 @SimenB I'm not happy with the current confidence level of "I doubt we'll add additional executables other than yarn and yarnpkg so I think it's fine to just hard-code those two." personally. 

What will happen for updates of Yarn is that someone will submit a PR to bump to the next version whenever it comes out. If Yarn ever does include additional executables, manpages, etc. then someone will need to notice and update the formula and until then Yarn and Homebrew both end up with a substandard situation.

This is why for similar e.g. C/C++ projects we strongly recommend having and using a `make install` so that upstream is control of what files get installed.

Thoughts?

**Comment by MikeMcQuaid** [comment]
> I would think that if other binaries are added, and the Debian script needs to be updated, someone™️ would remember to also make sure the linked binaries are correct.

I also just saw the Debian script. I think ideally there'd be a Unix install script (which the Debian one could call, even) which installed to a given `PREFIX`. If this logic lives in `yarn`'s repo: I have no issue with it 👍 

> I did have a solution that checked the bin entry in package.json, but we concluded that it was a bit too complex for little gain

Yeh, I appreciate you investigating that 🙇.

**Comment by ilovezfs** [comment]
>I think ideally there'd be a Unix install script (which the Debian one could call, even) which installed to a given PREFIX.

@Daniel15 that sure would be nice! :innocent:

File: Formula/yarn.rb
Comment (ilovezfs) [inline Formula/yarn.rb]: @SimenB let's add
```
revision 1
```

**Review by ilovezfs — CHANGES_REQUESTED** [review]

File: Formula/yarn.rb
Comment (SimenB) [inline Formula/yarn.rb]: Done. What about `mirror`?

**Review by ilovezfs — APPROVED** [review]

File: Formula/yarn.rb
Comment (ilovezfs) [inline Formula/yarn.rb]: I think we can go without for now since it's unlikely yarnpkg.com won't resolve

Comment (ilovezfs) [inline Formula/yarn.rb]: You'll need to move `revision` above `head` or it won't pass `brew audit --strict`

Comment (SimenB) [inline Formula/yarn.rb]: I'm currently getting this:
```
$ brew audit --strict yarn
Error: Error running `rubocop --format json --force-exclusion --config /usr/local/Homebrew/Library/.rubocop.yml /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/yarn.rb`
```

What am I doping wrong?

Comment (ilovezfs) [inline Formula/yarn.rb]: You can probably fix that by blowing away your `~/.gem` directory

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by ilovezfs** [comment]
Shipped! :rocket:

Thanks @SimenB for the PR and @Daniel15 for the help.

**Pull request merged**
